### ISSUE TITLE
Question re plotScores and pcs parameter

### ISSUE THREAD

**Issue by CCF49** [bryanhanson/ChemoSpec#18]
In your typical example "plotScores"  with parameter pcs=c(1,2)  creates a display plot of PC2 scores versus PC1 scores.  I assume that if pcs is changed to pcs=c(2,3) and the line executed that the new display plot would be PC3 scores versus PC2 scores.  However, when I make this change and execute the line the displayed plot remains identical to the original example (e.g. , PC2 vs. PC1).  Have I misunderstood the function of the pcs parameter? Any tips would be appreciated.  Thanks.

**Comment by bryanhanson** [comment]
Hi, that problem was introduced accidentally and was noticed and fixed back in October,

There is a new version that went to CRAN yesterday — 5.0.207 which will also load ChemoSpecUtils version 0.2.163.  Can you please install those and let me know it works?  Mostly bug fixes,  but a lot of internal stuff was changed to accommodate ChemoSpec2D, now awaiting CRAN approval.  So please let me how it work for you.

Bryan

**Comment by bryanhanson** [comment]
Since the new version has only been on CRAN for 24hrs, you might need to get it from the Austrian server, use `chooseCRANmirror`.

**Comment by CCF49** [comment]
Bryan,
  Thank you very much for your timely response.
  I made these updates using the package update feature in RStudio and ran
my code again. The issue with the pcs parameter appears to be resolved
based on performance of plotScores in one part of my code processing *classical
PCA*.  However, in another part of my code where* robust PCA *scores are
processed, plotScores is throwing a warning as follows:

*Warning message:*
*In if (!soOK) stop("Argument 'so' was not found or did not have class
prcomp or pcaCon") :*
*  the condition has length > 1 and only the first element will be used*

It looks like plotScores interprets classical and robust PCA scores
differently.  This wasn't an issue in the previous version.
I will investigate this further and provide you some details if I can't
resolve/understand the issue.  Any thoughts you may have would be
appreciated.
  In the meantime, please accept my compliments on a very nice tool.
Obviously, a great deal of work and thought has gone into this
development.  I also very much appreciate the willingness of your and your
colleagues to share your experience.
  Should you be interested, my application is not in analytical chemistry.
Rather, I am a biomedical engineer working in hearing mechanisms.  I'm
studying differences in the spectral mechanical impedance patterns
(wideband tympanograms) of middle ear mechanics following cochlear
implantation using PCA.  ChemoSpec is spot on for processing spectral
density functions.  I'm very curious in seeing what you have in your new
version.
  Thanks again.
Charlie

**Comment by CCF49** [comment]
Just saw this message.  When I performed my update, I confirmed the correct
version numbers were picked up.

**Comment by bryanhanson** [comment]
Charlie, thanks for the report.  It was a silly error on my part.  I introduced more systematic argument checking and I was not thinking about one aspect correctly.  I'm not sure how this escaped notice, I will have to think of how to catch this sort of thing in the future (it escaped not only me but the CRAN check farm).  The problem is actually in `ChemoSpecUtils`.  To update, please do:

```
install.packages("devtools")
library("devtools")
install_github(repo = "bryanhanson/ChemoSpecUtils@devel")

```
And see if it works for you.

Further background: I recently split some of `ChemoSpec` out into `ChemoSpecUtils` in order to support a new package, `ChemoSpec2D`.  This should have been invisible, but as you discovered, not quite!  Also, your observation that plotting scores was being handled differently is spot-on, you found me out!

Thanks for your kind words about the package.  Your application to tympanograms sounds quite interesting, it makes me happy to hear people using `ChemoSpec` in different ways.  It's been out for about 9 years now, and seems to have a pretty large user base.  I've been fortunate to correspond with lots of people doing very interesting projects.

Let me know if you run  into other problems.

Bryan

**Comment by CCF49** [comment]
Hi Bryan,
  Thanks very much for your highly responsive assistance!  Above and beyond
the call of duty!
  I installed the patch you prepared.  The installation went well.  After I
ran my code again, the previous warning with 'plotScores' processing robust
PCA scores had been eliminated.  However, at other points in my code I
started seeing similar warnings for 'plotScree', 'plotScores3D',
'plotScoresRGL', 'plotLoadings', and 'sPlotSpectra' when processing robust
PCA scores.  Everything appears to be fine when processing classical PCA
scores with these functions. 'hcaScores' appears to handle both classical
and robust PCA scores appropriately.  These are the only ChemoSpec
functions I have substantial experience with, so the issue may extend to
other functions.  I will likely have a some questions for you about the
tools for ANOVA-PCA, but that would be a diversion at this point.
  To illustrate the current issues to you I have attached a Word file
containing the RStudio Console Output.  You will see the warnings in RED on
Page 7 of 8.  I've annotated this file a bit to make it easy to read.
Basically the code does the following:
1. Sets up the R environment in RStudio
2. Reads in comma-delimited data using matrix2SpectraObject followed by
data checking and data summaries
3. Then a Classical PCA is performed followed by various plotting of
results.
4. Finally a Robust PCA is performed followed by similar plotting of
results.  *This is where the Warnings are thrown.*
  By providing this level of detail, I'm not asking you to debug my code.
I just thought this would be an efficient way to show you what I'm
experiencing.
  I hope you have a great weekend.  Please don't inconvenience yourself
over this.
Best regards,
  Charlie

**Comment by bryanhanson** [comment]
Ah Geez ... this is not my usual approach— to introduce more errors! I’ve been babysitting my grandkids for nearly a week, my brain must be turning to mush! I’ll look at it and get it right in the next 24 hrs. Sorry about that.

**Comment by CCF49** [comment]
Hey
 Priority on the grandkids! Those moments don't come often. Enjoy yourself.
 No problem with this other stuff. I understand exactly how this goes.
C

**Comment by bryanhanson** [comment]
Alright Charlie, if you don't mind when you have time, would you please get the devel version again, it is 0.2.197.  I am testing differently for errors and warnings and I don't see any.  I think I still need to fine tune some of the argument checking but it seems to work for the time being.  I'm so glad you asked this question.

**Comment by CCF49** [comment]
Bryan,
  Thank you for the revisions in DEV 0.2.197.  This version appears to
address the previous problems in plotting robust PCA scores.
  I do observe two issues:

1.  'plotScree' (perhaps formerly 'plotScree2') is not capable of handling
robust PCA scores. A console example follows:
          > plotScree(robust, main = mytitle)
          Error in UseMethod("plotScree") :
              no applicable method for 'plotScree' applied to an object of
class "c('conPCA', 'princomp')"
     This is not a new issue created in preparing 0.2.197.  It was among
the issues previously described.

2.  'plotScoresRGL' now handles robust PCA scores (correction of previous
known problem); however, 'plotScoresRGL' now loads and runs interactively
VERY slowly for BOTH classical and robust PCA scores (new behavior with
0.2.197).  Previously, in plotting classical PCA scores, it was highly
responsive interactively for rotations and zooms.  Now there is a 2-3
second delay before a plot change appears.  It seems there is a very long
computation loop that must run to completion before plot updates.  It could
be a memory leak.  On one occasion running 'plotScoresRGL' crashed RStudio.

Thanks,
   Charlie

**Comment by bryanhanson** [comment]
OK, please try ChemoSpecUtls devel branch 0.2.200  I have fixed the plotScree issue (that was really an oversight more than a bug).  I had seen that once before but forgot about it in the midst of the other troubleshooting, so thanks for the reminder.

The rgl issue: This works fine for me in the R GUI on a Mac, for either classical or robust scores.  Also fine in RStudio.  I have found rgl to be a little finicky.  Try re-installing it.  If you are on a Mac, you might try re-installing xquartz as well.  I don’t have a Windows machine to play with.

Let me know.  Bryan

**Comment by CCF49** [comment]
Bryan,
  Dev 0.2.200 appears to run well.  I continue to experience the
'plotScoresRGL' issues but have observed that this behavior is sporadic in
occurrence, occurring more often than not.  Various software resets and re
installations do not have a predictable effect so far.  I agree that it is
unlikely due to ChemoSpecUtils.
  Thanks very much for your assistance in aggressively addressing my
questions/issues once observed.  I look forward to your new ChemoSpec2D
release.
Charlie

**Comment by bryanhanson** [comment]
My pleasure Charile, and thanks  so much again for helping me identify these annoying mostly self-inflicted issues!
